### ISSUE TITLE
envoy getting killed by cgroup due to outofmemory

### ISSUE THREAD

**Issue by praneetpushpal** [envoyproxy/envoy#28986]
Recently we have been observing OOM on envoy and envoy gets restarted. 

**Envoy version: 1.25.8**

This is the configuration being used at envoyLB side:

```jinja
  - name: testlb
    address:
      socket_address:
        address: 0000:0000:0000:0000:0000:0000:0000:0000
        ipv4_compat: true
        port_value: 8080
        protocol: TCP
    filter_chains:
      transport_socket:
        name: envoy.transport_sockets.tls
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
          common_tls_context:
            tls_params:
              tls_minimum_protocol_version: "TLSv1_2"
            tls_certificates:
            - certificate_chain:
                filename: "/certs/cert.pem"
              private_key:
                filename: "/certs/key.pem"
              password:
                filename: "/etc/pkencryptkey/password"
            validation_context:
              trusted_ca:
                filename: "/certs/cacert.pem"
            alpn_protocols: h2,http/1.1
          require_client_certificate: true
      filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          codec_type: AUTO
          stat_prefix: testlb
          route_config:
            name: testlb
            virtual_hosts:
            - name: testlb
              domains:
              - "*"
              routes:
              - match:
                  prefix: "/"
                route:
                  cluster: testevent
                  timeout:
                    seconds: 1800
                decorator:
                  operation: testlb
              require_tls: NONE
          http_filters:
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
          server_name: envoy-ingress-testlb
          drain_timeout:
            seconds: 1800


  - name: testevent
    type: STRICT_DNS
    connect_timeout:
      seconds: 1800
    lb_policy: ROUND_ROBIN
    circuit_breakers:
      thresholds:
      - priority: DEFAULT
        max_connections: 2600
        max_requests: 2600
      - priority: HIGH
        max_connections: 2600
        max_requests: 2600
    load_assignment:
      cluster_name: eventreciver
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: testevent-headless.admin-ns.svc.cluster.local.
                port_value: 8080
    typed_extension_protocol_options:
      envoy.extensions.upstreams.http.v3.HttpProtocolOptions:
        "@type": type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions
        explicit_http_config:
          http2_protocol_options: {}
    transport_socket:
      name: envoy.transport_sockets.tls
      typed_config:
        "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
        common_tls_context:
          tls_params:
            tls_minimum_protocol_version: "TLSv1_2"
          tls_certificates:
          - certificate_chain:
              filename: "/certs/cert.pem"
            private_key:
              filename: "/certs/key.pem"
            password:
              filename: "/etc/pkencryptkey/password"
          validation_context:
            trusted_ca:
              filename: "/certs/cacert.pem"

``` 

As seen in the config we have circuitbreaker configured which is being hit consistently. This is the stats:
cluster.eventreciver.upstream_rq_pending_overflow: 21760
cluster.eventreciver.upstream_rq_pending_overflow: 21760
cluster.eventreciver.upstream_rq_pending_overflow: 22681
cluster.eventreciver.upstream_rq_pending_overflow: 23900
cluster.eventreciver.upstream_rq_pending_overflow: 23900
cluster.eventreciver.upstream_rq_pending_overflow: 23968
cluster.eventreciver.upstream_rq_pending_overflow: 23968
cluster.eventreciver.upstream_rq_pending_overflow: 23968
cluster.eventreciver.upstream_rq_pending_overflow: 24202
cluster.eventreciver.upstream_rq_pending_overflow: 24202
cluster.eventreciver.upstream_rq_pending_overflow: 24202
cluster.eventreciver.upstream_rq_pending_overflow: 24317
cluster.eventreciver.upstream_rq_pending_overflow: 24317
cluster.eventreciver.upstream_rq_pending_overflow: 24317
cluster.eventreciver.upstream_rq_pending_overflow: 25299
cluster.eventreciver.upstream_rq_pending_overflow: 26413
cluster.eventreciver.upstream_rq_pending_overflow: 26413
cluster.eventreciver.upstream_rq_pending_overflow: 26413

And, this shows that cgroup killed it 7 timesbecause of OOM:
**cat /sys/fs/cgroup/memory/memory.oom_control
oom_kill_disable 0
under_oom 0
oom_kill 7**

Can anyone please suggest what is going wrong here. I am verry new to Envoy, but these are my thoughts:
Is it that envoy is buffering the requests which are in overflow and ultimately goes out of memory? And, if that is the case can we have circuit_breaker configured at listeners section too such that buffering does not happen at envoy.
I know currently envoy does not have circuit_breaker  at listeners section, what else shall I look for to fix the issue.

**Comment by yanavlasov** [comment]
You may also need to configure overload manager: https://www.envoyproxy.io/docs/envoy/latest/configuration/operations/overload_manager/overload_manager.html

**Comment by praneetpushpal** [comment]
ok Thanks! I will try

**Comment by praneetpushpal** [comment]
I tried overload manager results are fine but, if requests are too frequent envoy still goes outofmemory.
However with flowcontrol the results are positive and envoy never goes OOM.
https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/flow_control


Just a little background of my issue:
Recently we migrated from envoyproxy/envoy-alpine:v1.21 to v1.25.8.  And, as we needed a smaller footprint we build the envoy binary from envoy source code on alpine 1.17. 

So, please suggest setting buffer limit is it a proper approach? And, as I dont want to break my production more can you please suggest important points that I should follow while setting this limit.
For now I have kept it like this:
per_connection_buffer_limit_bytes: 4096
initial_stream_window_size: 65535

**Comment by praneetpushpal** [comment]
> yanavlasov

I tried overload manager results are fine but, if requests are too frequent envoy still goes outofmemory.
However with flowcontrol the results are positive and envoy never goes OOM.
https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/flow_control

Just a little background of my issue:
Recently we migrated from envoyproxy/envoy-alpine:v1.21 to v1.25.8. And, as we needed a smaller footprint we build the envoy binary from envoy source code on alpine 1.17.

So, please suggest setting buffer limit is it a proper approach? And, as I dont want to break my production more can you please suggest important points that I should follow while setting this limit.
For now I have kept it like this:
per_connection_buffer_limit_bytes: 4096
initial_stream_window_size: 65535

**Comment by praneetpushpal** [comment]
Hi @alyssawilk 

I saw your post regarding flow control in multiple forums. So, tagging you. Sorry about that.
Need a clarification with flow control.

We are using envoy proxy and we have some upstream which are actually slow, sometimes taking upto 22 seconds:

[2023-10-23T08:37:01.659Z] "POST /test.sl.EventDB.v1.EventDB/Create HTTP/2" 200 - 529 7 *4006 4006* "-" "grpc-go/1.56.2" "77cab6b6-0638-430e-b59a-4c3b3029b3b5" "localhost:8887" "172.16.95.212:8080"
[2023-10-23T08:37:02.273Z] "POST /test.sl.EventDB.v1.EventDB/Create HTTP/2" 200 - 685 7 *3392 3392* "-" "grpc-go/1.56.2" "f5042724-1468-41a8-b6d5-75a5af0370b7" "localhost:8887" "172.16.95.212:8080"
[2023-10-23T08:36:42.770Z] "POST /test.sl.EventDB.v1.EventDB/Create HTTP/2" 200 - 498 7 ***22895 22895*** "-" "grpc-go/1.56.2" "5ae72910-3b33-4957-969d-731710bc3206" "localhost:8887" "172.16.95.212:8080"


In such a case envoy is buffering, resulting in out of memory.

In order to protect envoy we wanted to bring in flow control. The doubt is if we configure initial_stream_window_size and per_connection_buffer_limit_bytes to be less than the size of request, will envoy reject such requests?

For example: if I configure initial_stream_window_size  as 1 MB and my request is 2 MB will envoy reject such requests?

Basically I dont want envoy to buffer, it should just forward the request to upstream without rejecting requests. Please give your valuable suggestions

**Comment by alyssawilk** [comment]
Sorry for the delay on this - was out at EnvoyCon/KubeCon

if all your filters are streaming filters, your data should pass through no problem.  if you have hard-coded buffer limits _and_ you have buffering filters (like the buffer filter), then Envoy will reject requests / error on responses which go over the limits
